src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -192,27 +314,69 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_qp(struct rdma_cm_id *id,
     return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdamcm_cm_ep_set_qp_num(struct rdma_conn_param *conn_param,
-                            uct_rdmacm_cm_ep_t *cep)
+static ucs_status_t uct_rdamcm_cm_ep_create_qpn(uct_rdmacm_cm_ep_t *cep,
+                                                uint32_t* qpn)
 {
     struct ibv_cq *cq;
     ucs_status_t status;
+    bool supported;
 
-    status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs,
-                                  &cq);
+    status = uct_rdamcm_cm_ep_is_reserved_qpn_supported(cep, &supported);
     if (status != UCS_OK) {
         return status;
     }
 
-    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
-    status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
+    if (supported) {
+        status = uct_rdamcm_cm_ep_create_reserved_qpn(cep);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qpn;
+        }        
+    } else {
+        status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs, &cq);
+        if (status != UCS_OK) {
+            return status;
+        }
```

Comment:
due to `uct_rdamcm_cm_ep_is_reserved_qpn_supported` implementation, you are initializes 2 hash entries for 2 hashes for the same device, can we unite the contexts like this:
```
strict context {
 uint8_t flags; // <- support should be here
 uniton {
    struct {
...
    } reserved_qpn_context;
    struct {
...
    } dummy_qpn_context;
 };
};
```
and have only 1 hash table?

Reply:
@evgeny-leksikov Yes, we can. But don't feel having strong reason to do it because:
1. Even if use 2 hash entries, one of them will be empty, so it won't cost much more memory.
2. The "support" is only for Reserved QPN, it's not a concept of dummy QPN, so don't feel good to put it in the common part.
3. The hash of dummy QP is only for qps, is not really same concept as Reseved QPN context which includes mutex/blk_list/caps. We can treat qps of dummy QP as context, but no strong reason to do this.
4. Using above way will make the struct more complex/and the context members naming longer.
5. Mixing defination of Reserved QPN and dummy QPN will make things more conplext if later we want to use HAVE_DECL_MLX5DV_IS_SUPPORTED to compile out all Reserved QPN related code.

Reply:
@evgeny-leksikov Fixed.